src/monitoring/index.ts

```diff
@@ -0,0 +1,295 @@
+          .map(
+            (domain) =>
+              ({
+                domain,
+                type: IsomerHostedDomainType.REDIRECTION,
+              } as IsomerHostedDomain)
+          )
+      )
+  }
+
+  /**
+   * This is in charge of fetching all the domains that are are under Isomer, inclusive
+   * of any subdomains and redirects.
+   */
+  getAllDomains() {
+    this.monitoringServiceLogger.info("Fetching all domains")
+    return ResultAsync.combine([
+      this.getAmplifyDeployments().mapErr(
+        (error) => new MonitoringError(error.message)
+      ),
+      this.getRedirectionDomains(),
+      this.getKeyCdnDomains(),
+    ]).andThen(([amplifyDeployments, redirectionDomains, keyCdnDomains]) => {
+      this.monitoringServiceLogger.info("Fetched all domains")
+      return okAsync(
+        [...amplifyDeployments, ...redirectionDomains, ...keyCdnDomains].sort(
+          (a, b) => {
+            const domainA = a.domain
+            const domainB = b.domain
+            if (
+              domainA.startsWith("www.") &&
+              domainA.slice(`www.`.length) === domainB
+            ) {
+              return 0
+            }
+            if (
+              domainB.startsWith("www.") &&
+              domainA === domainB.slice(`www.`.length)
+            ) {
+              return 0
+            }
+            if (domainA === domainB) return 0
+            return domainA > domainB ? 1 : -1
+          }
+        )
+      )
+    })
+  }
+
+  // todo: once /siteup logic is merged into dev, we can add that as to alert isomer team
+  generateReportCard(domains: IsomerHostedDomain[]) {
+    const reportCard: ReportCard[] = []
+
+    const domainResolvers = domains.map(({ domain, type }) => {
+      const aRecord = ResultAsync.fromPromise(
+        dns.resolve(domain, "A"),
+        (e) => e
+      ).orElse(() => okAsync([]))
+      const quadArecord = ResultAsync.fromPromise(
+        dns.resolve(domain, "AAAA"),
+        (e) => e
+      ).orElse(() => okAsync([]))
+
+      const cNameRecord = ResultAsync.fromPromise(
+        dns.resolve(domain, "CNAME"),
+        (e) => e
+      ).orElse(() => okAsync([]))
+
+      const caaRecord = ResultAsync.fromPromise(
+        dns.resolve(domain, "CAA"),
+        (e) => e
+      )
+        .orElse(() => okAsync([]))
+        .map((records) => records.map((record) => record.toString()))
+
+      return ResultAsync.combineWithAllErrors([
+        aRecord,
+        quadArecord,
+        cNameRecord,
+        caaRecord,
+      ])
+        .andThen((resolvedDns) =>
+          okAsync<ReportCard>({
+            domain,
+            type,
+            aRecord: resolvedDns[0],
+            quadArecord: resolvedDns[1],
+            cNameRecord: resolvedDns[2],
+            caaRecord: resolvedDns[3],
+          })
+        )
+        .map((value) =>
+          reportCard.push({
+            ...value,
+          })
+        )
+        .andThen(() => okAsync(reportCard))
```

Comment:
hmm concretely what are you suggesting ah? if we just return we will return the array directly, we will get an empty array since we resolve before the promise resolves?

Reply:
Hmm, if we just return without the `map`, it'll be in a `ResultAsync` which we can ask downstream to consume isn't it? not sure if i'm misunderstanding you here

Reply:
wait ah `value` is of type `ReportCard`, but what we want is a list. 

The distinction to make here is that `ReportCard` is for one domain, and we want to map over every domains to get this.